urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/SevenZip/Compression/LZMA/Encoder.java

```diff
@@ -0,0 +1,1726 @@
+                                while (lenEnd < cur + offset)
+                                {
+                                    _optimum[++lenEnd].Price = kIfinityPrice;
+                                }
+                                curAndLenPrice = nextRepMatchPrice
+                                        + GetRepPrice(0, lenTest2, state2,
+                                                posStateNext);
+                                optimum = _optimum[cur + offset];
+                                if (curAndLenPrice < optimum.Price)
+                                {
+                                    optimum.Price = curAndLenPrice;
+                                    optimum.PosPrev = cur + lenTest + 1;
+                                    optimum.BackPrev = 0;
+                                    optimum.Prev1IsChar = true;
+                                    optimum.Prev2 = true;
+                                    optimum.PosPrev2 = cur;
+                                    optimum.BackPrev2 = curBack
+                                            + Base.kNumRepDistances;
+                                }
+                            }
+                        }
+                        offs += 2;
+                        if (offs == numDistancePairs)
+                        {
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+    
+    boolean ChangePair(int smallDist, int bigDist)
+    {
+        final int kDif = 7;
+        return smallDist < 1 << 32 - kDif && bigDist >= smallDist << kDif;
+    }
+    
+    void WriteEndMarker(int posState) throws IOException
+    {
+        if (!_writeEndMark)
+        {
+            return;
+        }
+        
+        _rangeEncoder.Encode(_isMatch, (_state << Base.kNumPosStatesBitsMax)
+                + posState, 1);
+        _rangeEncoder.Encode(_isRep, _state, 0);
+        _state = Base.StateUpdateMatch(_state);
+        final int len = Base.kMatchMinLen;
+        _lenEncoder.Encode(_rangeEncoder, len - Base.kMatchMinLen, posState);
+        final int posSlot = (1 << Base.kNumPosSlotBits) - 1;
+        final int lenToPosState = Base.GetLenToPosState(len);
+        _posSlotEncoder[lenToPosState].Encode(_rangeEncoder, posSlot);
+        final int footerBits = 30;
+        final int posReduced = (1 << footerBits) - 1;
+        _rangeEncoder.EncodeDirectBits(posReduced >> Base.kNumAlignBits,
+                footerBits - Base.kNumAlignBits);
+        _posAlignEncoder.ReverseEncode(_rangeEncoder, posReduced
+                & Base.kAlignMask);
+    }
+    
+    void Flush(int nowPos) throws IOException
+    {
+        ReleaseMFStream();
+        WriteEndMarker(nowPos & _posStateMask);
+        _rangeEncoder.FlushData();
+        _rangeEncoder.FlushStream();
+    }
+    
+    public void CodeOneBlock(long[] inSize, long[] outSize, boolean[] finished)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 103 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.